File: t/CS/PDFFormsDemo/PDFForms/PDFFormControls/TextBoxField.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.Controls;
using Leadtools.Pdf;

namespace PDFFormsDemo
{
   public class TextFormField : FormFieldControl
   {
      public TextFormField()
      {
         _pdfTextBox = new PDFTextBox();
         this.SetChildControl(_pdfTextBox);
      }

      public override void InitControl(PDFFormField formField)
      {
         base.InitControl(formField);

         TextBox textBox = _pdfTextBox.TextBox;

         foreach (string content in formField.Contents)
         {
            textBox.Text += content;
         }

         textBox.Text = textBox.Text.Replace("\n", "\r\n");

         InitTextBoxControl(textBox, formField);

         SetTextJustification(textBox, formField.TextJustification);

         if (this._autoFontResize)
         {
            textBox.TextChanged += PDFTextBox_TextChanged;
         }
      }

      public override void SetToolTip(string toolTip)
      {
         if (FormFieldControl.FormFieldsToolTip != null)
         {
            FormFieldControl.FormFieldsToolTip.SetToolTip(_pdfTextBox.TextBox, toolTip);
         }

         base.SetToolTip(toolTip);
      }

      public override void SetContextMenuStrip()
      {
         base.SetContextMenuStrip();

         if (FormFieldControl.FormFieldsContextMenu != null)
         {
            TextBox textBox = _pdfTextBox.TextBox;
            textBox.Tag = this;

            textBox.ContextMenuStrip = FormFieldControl.FormFieldsContextMenu;
         }
      }

      public override void UpdateFormField(PDFFormField formField)
      {
         base.UpdateFormField(formField);

         TextBox textBox = _pdfTextBox.TextBox;

         formField.Contents.Clear();

         formField.Contents.Add(textBox.Text);

         switch (textBox.TextAlign)
         {
            case HorizontalAlignment.Center:
               formField.TextJustification = PDFFormField.TextJustificationCentered;
               break;
            case HorizontalAlignment.Left:
               formField.TextJustification = PDFFormField.TextJustificationLeft;
               break;
            case HorizontalAlignment.Right:
               formField.TextJustification = PDFFormField.TextJustificationRight;
               break;
         }

         if (textBox.Multiline)
            formField.FieldFlags |= PDFFormField.FieldFlagsTextMultiline;
      }

      protected override void DoFontResize()
      {
         TextBox textBox = _pdfTextBox.TextBox;

         float fontSize = CalculateFontSize(textBox.Text, textBox.Multiline);

         _pdfTextBox.Font = new Font(_font.FontFamily, fontSize);
      }

      protected override void OnPropertyChange()
      {
         this._pdfTextBox.TextBox.ReadOnly = _isReadOnly;

         base.OnPropertyChange();
      }

      private void PDFTextBox_TextChanged(object sender, EventArgs e)
      {
         DoFontResize();
      }

      private void InitTextBoxControl(TextBox textBox, PDFFormField formField)
      {
         bool TextDoNotScroll = (formField.FieldFlags & PDFFormField.FieldFlagsTextDoNotScroll) == PDFFormField.FieldFlagsTextDoNotScroll;
         bool TextDoNotSpellcheck = (formField.FieldFlags & PDFFormField.FieldFlagsTextDoNotSpellcheck) == PDFFormField.FieldFlagsTextDoNotSpellcheck;
         bool TextMultiline = (formField.FieldFlags & PDFFormField.FieldFlagsTextMultiline) == PDFFormField.FieldFlagsTextMultiline;

         if (TextDoNotScroll)
            textBox.ScrollBars = ScrollBars.None;
         else
            textBox.ScrollBars = ScrollBars.Both;

         if (TextMultiline)
            textBox.Multiline = true;
         else
            textBox.Multiline = false;
      }

      private void SetTextJustification(TextBox textBox, int textJustification)
      {
         switch (textJustification)
         {
            case PDFFormField.TextJustificationLeft:
               textBox.TextAlign = HorizontalAlignment.Left;
               break;

            case PDFFormField.TextJustificationCentered:
               textBox.TextAlign = HorizontalAlignment.Center;
               break;

            case PDFFormField.TextJustificationRight:
               textBox.TextAlign = HorizontalAlignment.Right;
               break;
         }
      }

      #region Properties

      #region PDFTextBox

      private PDFTextBox _pdfTextBox;
      public PDFTextBox PDFTextBox
      {
         get { return this._pdfTextBox; }
      }

      #endregion //PDFTextBox

      #endregion //Properties
   }

   public class PDFTextBox : Panel, IFormFieldControl
   {
      private bool _focus = false;

      public PDFTextBox()
      {
         _textBox = new TextBox();
         _textBox.AcceptsReturn = true;
         _textBox.AutoSize = false;
         _textBox.BorderStyle = System.Windows.Forms.BorderStyle.None;
         _textBox.Dock = System.Windows.Forms.DockStyle.Fill;
         _textBox.MouseEnter += PDFTextBox_MouseEnter;
         _textBox.MouseLeave += PDFTextBox_MouseLeave;
         _textBox.GotFocus += _textBox_GotFocus;

         this.Controls.Add(this._textBox);
         this.MouseEnter += PDFTextBox_MouseEnter;
         this.MouseLeave += PDFTextBox_MouseLeave;
         this.Paint += PDFTextBox_Paint;
      }

      private void _textBox_GotFocus(object sender, EventArgs e)
      {
         ImageViewer imageViewer = (this.Parent.Parent as ImageViewer);
         if (imageViewer != null)
         {
            imageViewer.UpdateTransform();
         }
      }

      private void PDFTextBox_MouseEnter(object sender, EventArgs e)
      {
         _focus = true;
         this.Invalidate();
      }

      private void PDFTextBox_MouseLeave(object sender, EventArgs e)
      {
         _focus = false;
         this.Invalidate();
      }

      private void PDFTextBox_Paint(object sender, PaintEventArgs e)
      {
         UpdateControl();

         // Draw TextBox Border.
         FormFieldControl.DrawBorder(e.Graphics, this, this.ClientSize, _focus);
      }

      private void UpdateControl()
      {
         switch (_fieldBorderStyle)
         {
            case FieldBorderStyle.Solid:
            case FieldBorderStyle.Dashed:
               this.Padding = new Padding(_borderThickness);
               break;
            case FieldBorderStyle.Beveled:
            case FieldBorderStyle.Inset:
               this.Padding = new Padding(_borderThickness * 2);
               break;
            case FieldBorderStyle.Underlined:
               this.Padding = new Padding(0, 0, 0, _borderThickness);
               break;
         }
      }

      #region Properties

      #region ForeColor

      public override Color ForeColor
      {
         get
         {
            return _textBox.ForeColor;
         }
         set
         {
            _textBox.ForeColor = value;
         }
      }

      #endregion //ForeColor

      #region BackgroundColor

      protected Color _backgroundColor = SystemColors.Control;
      public Color BackgroundColor
      {
         get
         {
            return _backgroundColor;
         }
         set
         {
            _backgroundColor = value;
            _textBox.BackColor = this._backgroundColor;
         }
      }

      #endregion //BackgroundColor

      #region BorderColor

      protected Color _borderColor = Color.Black;
      public Color BorderColor
      {
         get
         {
            return _borderColor;
         }
         set
         {
            _borderColor = value;
         }
      }

      #endregion //BorderColor

      #region BorderThickness

      private int _borderThickness = 1;
      public int BorderThickness
      {
         get
         {
            return _borderThickness;
         }
         set
         {
            _borderThickness = value;
         }
      }

      #endregion //BorderThickness

      #region FieldBorderStyle

      protected FieldBorderStyle _fieldBorderStyle = FieldBorderStyle.Solid;
      public FieldBorderStyle FieldBorderStyle
      {
         get
         {
            return _fieldBorderStyle;
         }
         set
         {
            _fieldBorderStyle = value;
         }
      }

      #endregion //FieldBorderStyle

      #region FocusedBorderColor

      protected Color _focusedBorderColor = Color.FromArgb(255, 0, 153, 225);
      public Color FocusedBorderColor
      {
         get
         {
            return _focusedBorderColor;
         }
      }

      #endregion //FocusedBorderColor

      #region IsFieldVisible

      private bool _isFieldVisible = true;
      public bool IsFieldVisible
      {
         get
         {
            return _isFieldVisible;
         }
         set
         {
            _textBox.Visible = value;
            _isFieldVisible = value;
         }
      }

      #endregion //IsFieldVisible

      #region TextBox

      private TextBox _textBox;
      public TextBox TextBox
      {
         get { return _textBox; }
      }

      #endregion //TextBox

      #endregion //Properties
   }
}

File: t/VB/PDFFormsDemo/PDFForms/PDFFormControls/TextBoxField.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.Controls
Imports Leadtools.Pdf
Imports System

Namespace PDFFormsDemo
   Public Class TextFormField
      Inherits FormFieldControl
      Public Sub New()
         _pdfTextBox = New PDFTextBox()
         Me.SetChildControl(_pdfTextBox)
      End Sub

      Public Overrides Sub InitControl(formField As PDFFormField)
         MyBase.InitControl(formField)

         Dim textBox As TextBox = _pdfTextBox.TextBox

         For Each content As String In formField.Contents
            textBox.Text += content
         Next

         textBox.Text = textBox.Text.Replace(vbLf, vbCr & vbLf)

         InitTextBoxControl(textBox, formField)

         SetTextJustification(textBox, formField.TextJustification)

         If Me._autoFontResize Then
            AddHandler textBox.TextChanged, AddressOf PDFTextBox_TextChanged
         End If
      End Sub

      Public Overrides Sub SetToolTip(toolTip As String)
         If FormFieldControl.FormFieldsToolTip IsNot Nothing Then
            FormFieldControl.FormFieldsToolTip.SetToolTip(_pdfTextBox.TextBox, toolTip)
         End If

         MyBase.SetToolTip(toolTip)
      End Sub

      Public Overrides Sub SetContextMenuStrip()
         MyBase.SetContextMenuStrip()

         If FormFieldControl.FormFieldsContextMenu IsNot Nothing Then
            Dim textBox As TextBox = _pdfTextBox.TextBox
            textBox.Tag = Me

            textBox.ContextMenuStrip = FormFieldControl.FormFieldsContextMenu
         End If
      End Sub

      Public Overrides Sub UpdateFormField(formField As PDFFormField)
         MyBase.UpdateFormField(formField)

         Dim textBox As TextBox = _pdfTextBox.TextBox

         formField.Contents.Clear()

         formField.Contents.Add(textBox.Text)

         Select Case textBox.TextAlign
            Case HorizontalAlignment.Center
               formField.TextJustification = PDFFormField.TextJustificationCentered
               Exit Select
            Case HorizontalAlignment.Left
               formField.TextJustification = PDFFormField.TextJustificationLeft
               Exit Select
            Case HorizontalAlignment.Right
               formField.TextJustification = PDFFormField.TextJustificationRight
               Exit Select
         End Select

         If textBox.Multiline Then
            formField.FieldFlags = formField.FieldFlags Or PDFFormField.FieldFlagsTextMultiline
         End If
      End Sub

      Protected Overrides Sub DoFontResize()
         Dim textBox As TextBox = _pdfTextBox.TextBox

         Dim fontSize As Single = CalculateFontSize(textBox.Text, textBox.Multiline)

         _pdfTextBox.Font = New Font(_font.FontFamily, fontSize)
      End Sub

      Protected Overrides Sub OnPropertyChange()
         Me._pdfTextBox.TextBox.[ReadOnly] = _isReadOnly

         MyBase.OnPropertyChange()
      End Sub

      Private Sub PDFTextBox_TextChanged(sender As Object, e As EventArgs)
         DoFontResize()
      End Sub

      Private Sub InitTextBoxControl(textBox As TextBox, formField As PDFFormField)
         Dim TextDoNotScroll As Boolean = (formField.FieldFlags And PDFFormField.FieldFlagsTextDoNotScroll) = PDFFormField.FieldFlagsTextDoNotScroll
         Dim TextDoNotSpellcheck As Boolean = (formField.FieldFlags And PDFFormField.FieldFlagsTextDoNotSpellcheck) = PDFFormField.FieldFlagsTextDoNotSpellcheck
         Dim TextMultiline As Boolean = (formField.FieldFlags And PDFFormField.FieldFlagsTextMultiline) = PDFFormField.FieldFlagsTextMultiline

         If TextDoNotScroll Then
            textBox.ScrollBars = ScrollBars.None
         Else
            textBox.ScrollBars = ScrollBars.Both
         End If

         If TextMultiline Then
            textBox.Multiline = True
         Else
            textBox.Multiline = False
         End If
      End Sub

      Private Sub SetTextJustification(textBox As TextBox, textJustification As Integer)
         Select Case textJustification
            Case PDFFormField.TextJustificationLeft
               textBox.TextAlign = HorizontalAlignment.Left
               Exit Select

            Case PDFFormField.TextJustificationCentered
               textBox.TextAlign = HorizontalAlignment.Center
               Exit Select

            Case PDFFormField.TextJustificationRight
               textBox.TextAlign = HorizontalAlignment.Right
               Exit Select
         End Select
      End Sub

#Region "Properties"

#Region "PDFTextBox"

      Private _pdfTextBox As PDFTextBox
      Public ReadOnly Property PDFTextBox() As PDFTextBox
         Get
            Return Me._pdfTextBox
         End Get
      End Property

#End Region

#End Region
   End Class

   Public Class PDFTextBox
      Inherits Panel
      Implements IFormFieldControl
      Private _focus As Boolean = False

      Public Sub New()
         _textBox = New TextBox()
         _textBox.AcceptsReturn = True
         _textBox.AutoSize = False
         _textBox.BorderStyle = System.Windows.Forms.BorderStyle.None
         _textBox.Dock = System.Windows.Forms.DockStyle.Fill
         AddHandler _textBox.MouseEnter, AddressOf PDFTextBox_MouseEnter
         AddHandler _textBox.MouseLeave, AddressOf PDFTextBox_MouseLeave
         AddHandler _textBox.GotFocus, AddressOf _textBox_GotFocus

         Me.Controls.Add(Me._textBox)
         AddHandler Me.MouseEnter, AddressOf PDFTextBox_MouseEnter
         AddHandler Me.MouseLeave, AddressOf PDFTextBox_MouseLeave
         AddHandler Me.Paint, AddressOf PDFTextBox_Paint
      End Sub

      Private Sub _textBox_GotFocus(sender As Object, e As EventArgs)
         Dim imageViewer As ImageViewer = TryCast(Me.Parent.Parent, ImageViewer)
         If imageViewer IsNot Nothing Then
            imageViewer.UpdateTransform()
         End If
      End Sub

      Private Sub PDFTextBox_MouseEnter(sender As Object, e As EventArgs)
         _focus = True
         Me.Invalidate()
      End Sub

      Private Sub PDFTextBox_MouseLeave(sender As Object, e As EventArgs)
         _focus = False
         Me.Invalidate()
      End Sub

      Private Sub PDFTextBox_Paint(sender As Object, e As PaintEventArgs)
         UpdateControl()

         ' Draw TextBox Border.
         FormFieldControl.DrawBorder(e.Graphics, Me, Me.ClientSize, _focus)
      End Sub

      Private Sub UpdateControl()
         Select Case _fieldBorderStyle
            Case FieldBorderStyle.Solid, FieldBorderStyle.Dashed
               Me.Padding = New Padding(_borderThickness)
               Exit Select
            Case FieldBorderStyle.Beveled, FieldBorderStyle.Inset
               Me.Padding = New Padding(_borderThickness * 2)
               Exit Select
            Case FieldBorderStyle.Underlined
               Me.Padding = New Padding(0, 0, 0, _borderThickness)
               Exit Select
         End Select
      End Sub

#Region "Properties"

#Region "ForeColor"

      Public Overrides Property ForeColor() As Color
         Get
            Return _textBox.ForeColor
         End Get
         Set(value As Color)
            _textBox.ForeColor = value
         End Set
      End Property

#End Region

#Region "BackgroundColor"

      Protected _backgroundColor As Color = SystemColors.Control
      Public Property BackgroundColor() As Color Implements IFormFieldControl.BackgroundColor
         Get
            Return _backgroundColor
         End Get
         Set(value As Color)
            _backgroundColor = value
            _textBox.BackColor = Me._backgroundColor
         End Set
      End Property

#End Region

#Region "BorderColor"

      Protected _borderColor As Color = Color.Black
      Public Property BorderColor() As Color Implements IFormFieldControl.BorderColor
         Get
            Return _borderColor
         End Get
         Set(value As Color)
            _borderColor = value
         End Set
      End Property

#End Region

#Region "BorderThickness"

      Private _borderThickness As Integer = 1
      Public Property BorderThickness() As Integer Implements IFormFieldControl.BorderThickness
         Get
            Return _borderThickness
         End Get
         Set(value As Integer)
            _borderThickness = value
         End Set
      End Property

#End Region

#Region "FieldBorderStyle"

      Protected _fieldBorderStyle As FieldBorderStyle = FieldBorderStyle.Solid
      Public Property FieldBorderStyle() As FieldBorderStyle Implements IFormFieldControl.FieldBorderStyle
         Get
            Return _fieldBorderStyle
         End Get
         Set(value As FieldBorderStyle)
            _fieldBorderStyle = value
         End Set
      End Property

#End Region

#Region "FocusedBorderColor"

      Protected _focusedBorderColor As Color = Color.FromArgb(255, 0, 153, 225)
      Public ReadOnly Property FocusedBorderColor() As Color Implements IFormFieldControl.FocusedBorderColor
         Get
            Return _focusedBorderColor
         End Get
      End Property

#End Region

#Region "IsFieldVisible"

      Private _isFieldVisible As Boolean = True
      Public Property IsFieldVisible() As Boolean Implements IFormFieldControl.IsFieldVisible
         Get
            Return _isFieldVisible
         End Get
         Set(value As Boolean)
            _textBox.Visible = value
            _isFieldVisible = value
         End Set
      End Property

#End Region

#Region "TextBox"

      Private _textBox As TextBox
      Public ReadOnly Property TextBox() As TextBox
         Get
            Return _textBox
         End Get
      End Property

#End Region

#End Region
   End Class
End Namespace
